Declarative path

dan-rusu-inc opened the pull request:
Complex paths can be handled with Region sensors.  Given an ordered list of region sensors, when an entity enters a region, the motion of the entity is updated to navigate towards the next region in the list.

daniel-rusu committed:
Created utilities for handling entity IDs

daniel-rusu committed:
Declarative path using region sensors

Complex paths can be handled with Region sensors.  Given an ordered list
of region sensors, when an entity enters a region, the motion of the
entity is updated to navigate towards the next region in the list.

daniel-rusu committed:
Delete IDGenerator.java

daniel-rusu committed:
Delete DenseIDSet.java

daniel-rusu committed:
Delete IntStack.java

daniel-rusu committed:
Allow entities with different speeds to follow the path

davidrusu commented on src/gameengine/collisiondetection/tree/Quad.java:
```diff
@@ -260,8 +269,7 @@ public void childEntityUpdated(double timeToCheck, Entity entity) {
     public void relocateAndCheck(double timeToCheck, Entity entity) {
-        assert !isEntityInTree(entity) : "Entity should not be in the this tree when this method " +
-                "" + "is called";
+        assert !isEntityInTree(entity) : "Entity should not be in the this tree";
```
`in the this` -> `in this`

davidrusu commented on src/Utilities/GameUtils.java:
```diff
@@ -0,0 +1,90 @@
+package Utilities;
+
+import bricklets.Wall;
```
tight coupling with the bricklets package is not so nice, maybe move the wall class to `entities/common/` ?

dan-rusu-inc replied:
Agreed.  I felt ugly submitting that portion.  I planned to make the entity class accept handlers for updates & rendering so you no longer need to extend entity for trivial things (or maybe never) and then I'll delete the Wall class.

davidrusu commented on src/Utilities/LongQueue.java:
```diff
@@ -32,6 +32,8 @@ public LongQueue() {
      * @param initialCapacity The minimum initial capacity
      */
     public LongQueue(int initialCapacity) {
+        assert initialCapacity > 0;
```
👍

davidrusu commented on src/gameengine/entities/Entity.java:
```diff
@@ -21,7 +21,7 @@
     private int entityTypeBitMask = 1 << entityType;
     private Motion motion;
-    private Shape shape;
+    protected Shape shape;
```
why?

dan-rusu-inc replied:
I was too lazy to think when creating the Wall class.  Fixed back to private thanks.

davidrusu reviewed (approved):
Looks good besides a few minor comments

daniel-rusu committed:
Remove coupling from Shape to Entity and make shapes immutable

Cleanup to remove coupling from Shape to Entity and also made the shapes
immutable.  This also allows a shape to be used with more than 1 entity
at a time.

I rushed and was somewhat lazy so I just passed the entity references
when needed.  This could be made cleaner when we have a closer look at
it.

daniel-rusu committed:
Cleaning up the shape entities

Cleaning up the shape entity classes (in anticipation of removing the
need to extend entity just to work with a shape).
